stdlib/public/runtime/StaticBinaryELF.cpp

```diff
@@ -0,0 +1,354 @@
+//===--- StaticBinaryELF.cpp ----------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Parse a static ELF binary to implement lookupSymbol() address lookup.
+//
+//===----------------------------------------------------------------------===//
+
+#if defined(__ELF__) && defined(__linux__)
+
+#include "ImageInspection.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/StringRef.h"
+#include "swift/Basic/Lazy.h"
+#include <cassert>
+#include <string>
+#include <vector>
+#include <elf.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <linux/limits.h>
+#include <sys/auxv.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+using namespace std;
+using namespace llvm;
+
+#if __POINTER_WIDTH__ == 64
+#define ELFCLASS ELFCLASS64
+typedef Elf64_Ehdr Elf_Ehdr;
+typedef Elf64_Shdr Elf_Shdr;
+typedef Elf64_Phdr Elf_Phdr;
+typedef Elf64_Addr Elf_Addr;
+typedef Elf64_Word Elf_Word;
+typedef Elf64_Sym Elf_Sym;
+typedef Elf64_Section Elf_Section;
+#define ELF_ST_TYPE(x) ELF64_ST_TYPE(x)
+#else
+#define ELFCLASS ELFCLASS32
+typedef Elf32_Ehdr Elf_Ehdr;
+typedef Elf32_Shdr Elf_Shdr;
+typedef Elf32_Phdr Elf_Phdr;
+typedef Elf32_Addr Elf_Addr;
+typedef Elf32_Word Elf_Word;
+typedef Elf32_Sym Elf_Sym;
+typedef Elf32_Section Elf_Section;
+#define ELF_ST_TYPE(x) ELF32_ST_TYPE(x)
+#endif
+
+extern const Elf_Ehdr elfHeader asm("__ehdr_start");
+
+// Create strong linkage to pthread_self, pthread_once and pthread_key_create
+// as they are normally weak-linked and used to detect the presence of pthreads.
+// Without this the calls just jump to 0x0.
+__attribute__((__visibility__("hidden")))
+ pthread_t (*__strong_pthread_self)(void) = pthread_self;
+
+__attribute__((__visibility__("hidden")))
+int (*__strong_pthread_once)(pthread_once_t *, void (*)(void)) = pthread_once;
+
+__attribute__((__visibility__("hidden")))
+int (*__strong_pthread_key_create)(pthread_key_t *, void (*)(void *)) = pthread_key_create;
+
+
+class StaticBinaryELF {
+
+private:
+  // mmap a section of a file that might not be page aligned.
+  class Mapping {
+  public:
+    void *mapping;
+    size_t mapLength;
+    off_t diff;
+
+    Mapping(int fd, size_t fileSize, off_t offset, size_t length) {
+      if (fd < 0 || offset + length > fileSize) {
```

Comment:
Given that this is internal, perhaps make this a fatal error?  The user should validate that the `fd` and `size` are valid before creating a mapping.

Reply:
I made this into an `assert` as that was used elsewhere in the file.